glean-core/src/lib.rs

```diff
@@ -514,37 +563,7 @@ pub fn shutdown() {
         log::error!("Can't shutdown dispatcher thread: {:?}", e);
     }
 
-    // Call on_shutdown.
-    // Need to be done without holding a lock on the global Glean object.
-    {
-        let (tx, rx) = unbounded();
-
-        let handle = thread::Builder::new()
-            .name("glean.shutdown".to_string())
-            .spawn(move || {
-                let state = global_state().lock().unwrap();
-                state.callbacks.shutdown();
-
-                // Best-effort sending. The other side might have timed out already.
-                let _ = tx.send(()).ok();
-            })
-            .expect("Unable to spawn thread to wait on shutdown");
-
-        // TODO: 30 seconds? What's a good default here? Should this be configurable?
-        // Reasoning:
-        //   * If we shut down early we might still be processing pending pings.
-        //     In this case we wait at most 3 times for 1s = 3s before we upload.
-        //   * If we're rate-limited the uploader sleeps for up to 60s.
-        //     Thus waiting 30s will rarely allow another upload.
-        //   * We don't know how long uploads take until we get data from bug 1814592.
-        let result = rx.recv_timeout(Duration::from_secs(30));
-
-        if result.is_err() {
-            log::warn!("Waiting for upload failed. We're shutting down.");
-        } else {
-            let _ = handle.join().ok();
-        }
-    }
+    uploader_shutdown();
```

Comment:
Musing: I was reading through `uploader_shutdown` wondering if we could indeed call `with_glean` safely, and then get out here and read up-method to find that we do guard against init being fully complete, so that's helpful... but oof, this shutdown stuff is getting more and more complex and I'm not sure where the ideal place is to document or manage this complexity.

Reply:
ugh, yeah, it's ... a lot and spread in several places. This could definitely use some refactoring to streamline it again.